### DISCUSSION TITLE
Deep Aggregation its possible?

### DISCUSSION THREAD

**Discussion by Enhed** [directus/directus#9808] [Archive]
i can't create request for deep count or any aggregations. For example i just get o2m values like without "deep" query
![image](https://user-images.githubusercontent.com/36981278/141798088-325bc33b-2ec1-48b6-b463-84ddaa90776c.png)

and if i delete `_` from count i get error:
![image](https://user-images.githubusercontent.com/36981278/141798144-50e8ca17-5512-49c7-98b9-78026239b62c.png)

**Comment by azrikahar** [comment]
It is indeed possible! I believe this is just a case of improper syntax. You should do `[_aggregate][count]` instead of `[_aggregate[count]]` 👍

**Comment by Enhed** [comment]
i got another error:
![image](https://user-images.githubusercontent.com/36981278/141952844-6f9352c7-6c7b-439c-aec0-03b85a5a0855.png)

**Reply by azrikahar** [reply]
May I know what is the Directus version you're using? I just did a quick test and it seems to work:

Version: 9.0.1
Relationship: `units` O2M `cameras`

## Without deep aggregate

![image](https://user-images.githubusercontent.com/42867097/141959564-112d7b5f-9c70-4a99-bcd8-e39e19e1d6f5.png)

## With deep aggregate

![image](https://user-images.githubusercontent.com/42867097/141959496-2328d21c-b775-4a0f-8c77-b9bad6f3d295.png)

**Comment by Enhed** [comment]
ver. 9.0.0-rc.97, ill update them and try again, thanks you

**Reply by Enhed** [reply]
not working for me on latest, for example i want to get users count in role ( #9824 ) but get error:\
![image](https://user-images.githubusercontent.com/36981278/142165003-600daa19-d521-40a1-9726-b754b24407b8.png)

**Reply by wc-matteo** [reply]
same error

**Reply by ahmtaras** [reply]
I am also getting the same error. I asked on the Discord server but got no response.
The version I am using: 9.0.1

**Reply by mlbiche** [reply]
It is still not working, isn't it ? I am getting the same error related to the GROUP BY clause.

@azrikahar Are still able to run the request you screenshot ?

Maybe mine is not working because it is an M2M relation ?

**Reply by azrikahar** [reply]
@mlbiche I believe you might be facing #10047, where it doesn't readily work for Postgres/MySQL. Are you using one of them by any chance? (just so we can confirm this is unique to those)

In the mean time, to elaborate further what was the issue in #10047, a setup such as:

- collection `a`
  - primary key field `id`
  - string field `name`
  - m2m field `m2m_b`
- collection `b`
  - primary key field `id`
  - string field `name`
  
will have a junction table:

- collection `a_b`
  - primary key field `id`
  - foreign key field `a_id`
  - foreign key field `b_id`
  
When we are querying a deep aggregate count on `m2m_b` in collection `a`, the following request works for SQLite: `http://localhost:8055/items/a?deep[m2m_b][_aggregate][count]=id`

However for Postgres/MySQL, we will need to ensure the generated sql query's `GROUP BY` (_groupBy) and `ORDER BY` (sort) targets the same field as described in #10050, so the request will have to be: `http://localhost:8055/items/a?deep[m2m_b][_aggregate][count]=id&deep[m2m_b][_groupBy][]=a_id&deep[m2m_b][_sort]=a_id`

To make the above request easier to understand, here's how it will look like when using json or the SDK:

```js
deep: {
	m2m_b: {
		_aggregate: { count: 'id' },
		_groupBy: ['a_id'],
		_sort: 'a_id',
	},
},
```


Can you try the above suggestion and see whether it works for you?

**Reply by mlbiche** [reply]
I confirm that we are using Postgres 👍 

I tried out your solution and it works properly, thank you 🎉 

Is there a GitHub issue related to the Postgres/MySQL specific issue, so I can follow its evolution?

**Reply by azrikahar** [reply]
> I confirm that we are using Postgres 👍
>
> I tried out your solution and it works properly, thank you 🎉

Thanks for confirming!

> Is there a GitHub issue related to the Postgres/MySQL specific issue, so I can follow its evolution?

I don't recall that there's an existing GitHub issue for this, but I believe this may actually be more of a prerequisite knowledge with aggregations in general (regardless of db vendor) as it is technically correct, which is why said db vendors requires us to form the correct SQL query. 

That being said, there's definitely room for improvement from the API side to maybe return a better error along the lines of "group and sort should be included when aggregating" for example, or maybe some guidance in the docs on the necessity of group and sort in terms of the correctness of the formed SQL query under the hood.

**Reply by mlbiche** [reply]
Yes, I think the docs should state this specificity. I'll try to propose a merge request 👍

**Reply by rijkvanzanten** [reply]
> I don't recall that there's an existing GitHub issue for this, but I believe this may actually be more of a prerequisite knowledge with aggregations in general (regardless of db vendor) as it is technically correct, which is why said db vendors requires us to form the correct SQL query.

@azrikahar Lets track that as a bug. You shouldn't be required to know the underlying sql magic for this particular query style. If sort and group are always required to be configured to the same field, we could maybe enforce that in the API logic?

**Reply by azrikahar** [reply]
Tracking this in #17142 👍
